ocaml/idl/datamodel_vtpm.ml

```diff
@@ -39,24 +67,29 @@ let t =
       [
         (Published, rel_rio, "Added VTPM stub")
       ; (Extended, rel_next, "Added ability to manipulate contents")
-      ; (Extended, rel_next, "Added VTPM profiles")
-      ; (Changed, rel_next, "Removed backend field")
+      ; (Extended, rel_next, "Added VTPM unique and protected properties")
+      ; (Extended, rel_next, "Added Persistence backed")
       ]
-    ~gen_constructor_destructor:true ~name:_vtpm ~descr:"A virtual TPM device"
-    ~gen_events:false ~doccomments:[]
+    ~gen_constructor_destructor:false ~name:_vtpm ~descr:"A virtual TPM device"
+    ~gen_events:true ~doccomments:[]
     ~messages_default_allowed_roles:_R_POOL_ADMIN
     ~contents:
       [
         uid _vtpm
       ; field ~qualifier:StaticRO ~ty:(Ref _vm) "VM"
           "The virtual machine the TPM is attached to"
-      ; field ~qualifier:DynamicRO
-          ~ty:(Map (String, String))
-          ~lifecycle:[(Published, rel_next, "Added VTPM profiles")]
-          "profile" "The security properties that define how the TPM is handled"
+      ; field ~qualifier:DynamicRO ~ty:(Ref _vm) "backend"
+          "The domain where the backend is located (unused)"
+      ; field ~qualifier:DynamicRO ~ty:persistence_backend ~lifecycle:[]
+          "persistence_backend" "The backend where the vTPM is persisted"
+      ; field ~qualifier:StaticRO ~ty:Bool ~default_value:(Some (VBool false))
+          ~lifecycle:[] "unique"
+          "Whether the contents are never copied, satisfying the TPM spec"
+      ; field ~qualifier:DynamicRO ~ty:Bool ~lifecycle:[] "protected"
```

Comment:
There is code in XAPI that would check (see dm_api.ml) that all fields that are not in `rel_rio` have default values, which are required for DB upgrades to work.
However because the object is `rel_rio` I think all fields inherit that. But none of these new fields existed in `rel_rio`, so I think we need to add an explicit `in_product_since` definition to each field that we've added since Yangtze.

Reply:
This can be fixed in datamodel itself by setting in_product_since to None if lifecycle is None, so we don't need to do anything about Rio here for now.